### ISSUE TITLE
State Store Query API

### ISSUE THREAD

**Issue by dmitsh** [dapr/dapr#3662]
## In what area(s)?

/area runtime
/area docs

## Describe the proposal
This proposal is a follow-up on #1339 

This is a proposed design of the API for querying the **state store**.

:exclamation: ***Note:*** The purpose of this API is to provide the means for querying the state in a **state store**. This is **not** a generic database query API. 

This design is split into 3 independent parts:
1. Syntax of the query request/response
2. Intermediate query representation
3. Interface for constructing final native query in the state component

These parts could be changed/updated independently of each other, providing flexibility in further development.
Below is mode detailed description.

### 1.  Syntax of the query request

A user submits a query request via HTTP POST or gRPC.
The body of the request is the JSON map with 3 entries: `filter`, `sort`, and `pagination`
The `filter` section specifies the query conditions in the form of a tree of key/values operations, where the key is the operator and the value is the operands.

Currently we support four operations:
| Operator | Operands | Description |
|----------|-----------|-------------|
| `EQ`       | key:value  | key == value|
| `IN`        | key:[]value| key == value[0] OR key == value[1] OR ... OR key == value[n] | 
| `AND`    | []operation|  operation[0] AND operation[1] AND ... AND operation[n] |
| `OR`      | []operation|  operation[0] OR operation[1] OR ... OR operation[n] |

This design allows to add more operations without breaking changes.

The `sort` is an optional section is an ordered array of `key:order` pairs, where `key` is a key in the state store, and the `order` is an optional string indicating sorting order: `"ASC"` for ascending and `"DESC"` for descending. If omitted, ascending order is assumed.

The `pagination` is an optional section containing `limit` and `token` parameters. `limit` sets the page size. `token` is an iteration token returned by the component, and used in consecutive queries.

```go
type QueryRequest struct {
	Query Query `json:"query"`
}

type Query struct {
	Filters map[string]interface{} `json:"filter"`
	Sort    []Sorting              `json:"sort"`
	Page    Pagination             `json:"page"`

	// derived from Filters
	Filter Filter
}

type Sorting struct {
	Key   string `json:"key"`
	Order string `json:"order,omitempty"`
}

type Pagination struct {
	Limit int    `json:"limit"`
	Token string `json:"token,omitempty"`
}
```

Here is an example of the query request JSON body:
```json
{
    "query": {
        "filter": {
            "OR": [
                {
                    "EQ": {"person.org": "Dev Ops"}
                },
                {
                    "AND": [
                        {
                            "EQ": {"person.org": "Finance"}
                        },
                        {
                            "IN": {"state": ["CA", "WA"]}
                        }
                    ]
                }
            ]
        },
        "sort": [
            {
               "key": "state",
                "order": "DESC"
            },
            {
                "key": "person.name"
            }
        ],
        "pagination": {
            "limit": 10
        }
    }
}
```

The query request will be eventually translated into a native query language and executed in the state store component.

In the case of SQL, the aforementioned query will be translated into
```sql
SELECT * FROM c WHERE
  person.org = "Dev Ops" OR
  (person.org = "Finance" AND state IN ("CA", "WA"))
ORDER BY
  state DESC,
  person.name ASC
LIMIT 10
```

Upon successful execution, the component will return a JSON object with query results and pagination token:
```go
type QueryResponse struct {
	Results []QueryResult `json:"results"`
	Token   string        `json:"token,omitempty"`
}

type QueryResult struct {
	Key   string  `json:"key"`
	Data  []byte  `json:"data"`
	ETag  *string `json:"etag,omitempty"`
	Error string  `json:"error,omitempty"`
}
```

### 2. Intermediate query representation

Upon receiving the query request, Dapr will validate it and transform into Intermediate query object `MidQuery`,
which will be passed on to the state store component.

To leverage JSON query format, `MidQuery` uses custom JSON marshaling to populate tree of `Filter`s.
 
```go
type Filter interface {
	Parse(interface{}) error
}

type EQ struct {
	Key string
	Val interface{}
}

type IN struct {
	Key  string
	Vals []interface{}
}

type AND struct {
	Filters []Filter
}

type OR struct {
	Filters []Filter
}
```
### 3. An interface for constructing final native query in the state component

To simplify the process of query translation, we propose using [visitor design pattern](https://datacadamia.com/data/type/tree/visitor). A state store component developer would need to implement the `visit` method, and the runtime will use it to construct the native query statement.

```go
type Visitor interface {
	// returns "equal" expression
	VisitEQ(*EQ) (string, error)
	// returns "in" expression
	VisitIN(*IN) (string, error)
	// returns "and" expression
	VisitAND(*AND) (string, error)
	// returns "or" expression
	VisitOR(*OR) (string, error)
	// receives concatenated filters and finalizes the native query
	Finalize(string, *Query) error
}
```

The Dapr runtime implements `QueryBuilder` object that takes in `Visitor` interface and constructs the native query.

```go
type QueryBuilder struct {
	visitor Visitor
}

func (h *QueryBuilder) BuildQuery(q *Query) error {...}
```

The last part is to implement `Querier` interface in the component:
```go
type Querier interface {
	Query(req *QueryRequest) (*QueryResponse, error)
}
```

A sample implementation might look like that:
```go
func (m *MyComponent) Query(req *state.QueryRequest) (*state.QueryRespond, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	q := &Query{} // Query implements Visitor interface
	qbuilder := query.NewQueryBuilder(q)
	if err := qbuilder.BuildQuery(&req.Query); err != nil {
		return &state.QueryRespond{}, err
	}
	res, token, err := q.execute(ctx)
	if err != nil {
		return &state.QueryRespond{}, err
	}
	return &state.QueryRespond{
		Results:  res,
		Token: token,
	}, nil
}
```

**Comment by yaron2** [comment]
Thanks, generally this looks good.

The one concern I have is with the using facing `.filter` mechanism. It might be a tad overly simplistic. I'd like to consider ANSI SQL instead with clear definitions on which constructs are supported. The benefits of this are:

1. A known SQL query
2. Ability to expand to more SQL features over time (not doable with the current implementation).

wdyt?

**Comment by jjcollinge** [comment]
Thanks for the initial proposal - I need some time to form my thoughts but one thing I'd like to raise upfront would be the need for a way to `select` what columns/properties are to be return rather than assuming all. Appreciate we want to narrow the scope as much as possible here but I'd think this would be a required feature for any column-oriented databases.

**Comment by CodeMonkeyLeet** [comment]
@dmitsh a couple of clarifications on the proposal:

- Is this API intended to only ever provided filtering functionality? (i.e. supporting `LIKE` or `OR`, or even more general ANSI SQL query semantics would be implemented on a different non-break API in the future). The encoding of the query request into separate structures suggests it would be hard to extend the supported syntax/query features without also modifying the structures, which would be API breaking.
   - If only for filtering, then maybe renaming it as the State Store Filtering API is more appropriate, and leaves open the option for a different Query API in the future.

- Can you elaborate on the `visitor` pattern you plan to implement? It's not obvious from the proposal what the unit of iteration  that the component needs to operate on in the `visit` function. (i.e. what is passed into `visit` on each call?)

- How does the user page through the results? Presumably this would involve the `Pagination.token` property, but it's not clear how that's intended to be used.

**Comment by daixiang0** [comment]
My concern is that modern SQL servers use their own protocols to achieve high performance of SQL executions, after this there may be a loss for performance, wdyt? @dmitsh

**Comment by artursouza** [comment]
SQL query is beyond what a generic state store can process (JOIN, for example). I like that we start with a simple expression tree. DynamoDB started that way and recently introduced PartiQL as a custom built query language that matches the feature set of the database.

I think we should consider adding a query language later on but not in this iteration.

**Comment by dmitsh** [comment]
Thank you all for reviewing this proposal. This is still work in progress, and I will keep updating the content.
Since `GitHub issues` does not support revisions, I will do my best to make localized/focused changes at a time.
Per @CodeMonkeyLeet suggestion, I renamed it to "State Store Query API".
Also, I updated the introduction and the first section "Syntax of the query request".
Please review and comment.
These changes partially address comments from @yaron2 and @CodeMonkeyLeet .
Another clarification is that this API exclusively targets the State Store components.
We have another work item for the database query API, which should address comments from @jjcollinge and @daixiang0 .

I'm planning to update sections 2 and 3 in the upcoming days. Once that's done, I will remove "(WIP)" suffix from the title.

**Comment by artursouza** [comment]
By looking at the PR, I noticed that it does not handle 2 scenarios:

1. Prepared statements: to avoid SQL injection at the component level, the Visitor interface should also return the map of key-value pairs. Does CosmosDB support prepared statement in their client interface?
2. Native filter object: by looking at the MongoDB implementation, it feels like the filter object could be generated via the Visitor interface instead of deserializing it from the string. If the Visitor returns an interface{} instead of string, then each component can generate the object that makes more sense for its native client. It can be a string or a struct, for example.

**Comment by dmitsh** [comment]
@artursouza , how do you envision HTTP API for prepared statements? Will both the statement and the parameters be passed as part of the HTTP request? If so, this won't mitigate SQL injection.

**Comment by artursouza** [comment]
The HTTP API does not have this problem. This is about the query to the database.

**Comment by dmitsh** [comment]
> The HTTP API does not have this problem. This is about the query to the database.

I'm not designing an API to a database. I'm designing an API that provides a generic way to send a query to  any statestore component.
The statestore component, in turn, will have to implement this API and leverage an existing native database API to run the query.

In my PR https://github.com/dapr/components-contrib/pull/1153 I have added querying functionality to `cosmosdb` and `mongodb`. The `cosmosdb` API uses prepared statement, while `mongodb` API does not. 
This API works with both models.

If this answer does resolve your comment, please elaborate in details, what threat do you see in ***this*** API, and what is your suggestion to mitigate it.

**Comment by artursouza** [comment]
Let me rephrase my feedback: this is not about the API it is about how the query is constructed per database, then at that point - it is a database API. This is about how the code *uses* the database API and not about Dapr's query API. This can be addressed without changing anything in the API proposed, only in the implementation on how the query for each database is constructed.

The query to CosmosDB does not used a prepared statement. In CosmosDB, it is called parameterized query: https://docs.microsoft.com/en-us/azure/cosmos-db/sql/sql-query-parameterized-queries#examples See line 140 in cosmosdb_query.go and you will see that your code does not use the `parameters` attribute since all the values are embedded in the string. My feedback is that the output of the Builder should not only be the query but the parameters of the query.

Now by looking at the MongoDB implementation, the JSOn object for the query is build as String and not as a struct type, so some specific values of "key" or "value" could generate an invalid JSON or even alter the query to return a different result. For example, return values from the whole database instead just for the current user in the app. This, of course, assumes that `value` is coming from an user-input on the app, which can happen. So, this can be mitigated by constructing the query struct directly and not a string for later deserialization. After, we should also consider if there is a need to sanitize the values to avoid injection attacks: https://severalnines.com/database-blog/securing-mongodb-external-injection-attacks

I hope it clarifies my feedback.

If more clarification is needed, I see two options: wait for more people to give feedback, so having multiple points of view on this problem might help create clarity; or wait for me (or someone from the community - anyone?) to have a draft PR on top of this one showing how possible malicious values can temper the query on CosmosDB and MongoDB.

**Comment by dmitsh** [comment]
I understand the concept and risk of query injection. However, in this API both the query and the parameters are supplied by the user. If a bad actor can gain access to the API endpoint, they could change the query as easily as the parameters. So, the parametrized query won't help.

**Comment by artursouza** [comment]
> > The query to CosmosDB does not use a prepared statement. In CosmosDB, it is called parameterized query: https://docs.microsoft.com/en-us/azure/cosmos-db/sql/sql-query-parameterized-queries#examples See line 140 in cosmosdb_query.go and you will see that your code does not use the `parameters` attribute since all the values are embedded in the string. My feedback is that the output of the Builder should not only be the query but also the parameters of the query
> 
> I actually do use parametrized query in `CosmosDB` implementation of the visitor interface. Take a look at the function `setNextParameter`:
> 
> ```go
> func (q *Query) setNextParameter(val string) string {
> 	pname := fmt.Sprintf("@__param__%d__", len(q.query.Parameters))
> 	q.query.Parameters = append(q.query.Parameters, documentdb.Parameter{Name: pname, Value: val})
> 
> 	return pname
> }
> ```

Thanks for clarifying that and sorry for missing that part. I will come up with a PR to explain my point for the MongoDB part of the code. Great work here! This is looking good.
